### DISCUSSION TITLE
Grammar to describe a behavior tree,or other patterns?

### DISCUSSION THREAD

**Discussion by Qrnbth** [terrastruct/d2#987] [General]
Sometimes state machine is quite unclear for understanding, a tree of actions that indecate possible moves is quite helpful when a traverse to all the path of a set of thing needs to be done.

![image](https://user-images.githubusercontent.com/54792110/223305897-41b8a0fc-4b5b-4b32-b908-df6c753caf34.png)

like a simple business logic above (made with obsidian canvas), it has parts that repeats conditionaly,
if i m gonna describe a diagram with d2 like a tree, it could be something like this below:

![image](https://user-images.githubusercontent.com/54792110/223312669-f96668e9-caa8-47d6-b186-f6eebf96352e.png)

That's a simple logic already, but i could still get mistake or missing parts,
if it comes with more complexity,copy n paste and couter marking are gonna be annoying jobs, and it's hard to do some refactoring 

I'd checked around, but no such a product or grammer found designed for this,
conditional nodes generation, node alias for duplication, and recursive constraints might be helpful to acheive this,
but grammer design and implemetation are just way beyond my capbility, any idea?

**Comment by tmc** [comment]
You might consider a Statechart for this type of modeling. I started a discussion at https://github.com/terrastruct/d2/discussions/1128

**Reply by Qrnbth** [reply]
That might also work though,as they are quite similar problems,
but I'm actually wanted to discuss for the ability to describe specific rules then expand them into some graph.
It can be more expressive I think, but clearly there can be barely have a chance to have any sort of implementation...
Another approach is nice anyway
